### DISCUSSION TITLE
HD63B03RP Address mode?

### DISCUSSION THREAD

**Discussion by TheEccentricGenius** [KenWillmott/HD6303EVB#8] [General]
Which address mode would you advise?

I am leaning to getting a 74LS373 and doing it on the Address/Data lines and would you get the HD14053B for mode selection?

I'm only leaning towards the 74LS373 to get extra pins.

Still working on a basic schematic....

**Comment by KenWillmott** [comment]
I would like to use expanded multiplexed mode with an 74HC373, also to satisfy the multiplexed interface requirements of the HD146818 RTC IC. I had a look at the mode pins. Although there are three inputs, The 6303R version only implements modes 1,2, and 4. The expanded mode is selected whenever P20 is LOW at reset time, so tying it to ground with a resistor will do that. The other two mode select pins are "don't care" in this case. It is only the 6303V1 parts that implement all 8 modes.

P20 connects only with the input capture hardware, so I'm willing to sacrifice it for now, if the input capture function is needed, it could be dealt with, using some additional add on circuit. It has no effect on the serial interface, so I'm not concerned about it.

It might be a good idea to provide PCB trace or pin jumpers for that pin, to make it easy to modify if needed.

**Reply by TheEccentricGenius** [reply]
So, Pins 11 & 12 are used for serial right?
I'll just do the resistor idea and provide a jumper for easy modification.

Will a 8MHz crystal do or should I consider acquiring a 7.3728MHz crystal(Only asking because the current board has the latter)?

**Comment by KenWillmott** [comment]
An 8MHz crystal gives no more than 4800 baud with the internal divider. For 115200, you need a 7.3728MHz crystal, or an external 1.8432MHz clock on pin 10.

**Reply by TheEccentricGenius** [reply]
I am getting the 7.3728MHz crystal as recommended.
Pin compatible so far(Not used pins left empty only added AS, which can go on pin 23 or 29 in the header?).

Do you think a 74HC139 would work as glue logic for the missing WE & OE lines?

Here is my schematic so far.
![image](https://github.com/KenWillmott/HD6303EVB/assets/137060094/2c4e87bd-1c81-47da-8415-c1b00ccf80aa)

**Comment by KenWillmott** [comment]
AS should go to pin 29 (Q) on the expansion header, because the timing is similar enough that some boards might use it. The problem of defining bus signal alternatives for non-6809 processors on the M8 bus is a "to do". I would like to make a matrix chart that defines the signal mapping but for now, I believe the AS signal is the only one that differentiates the 6303Y and 6303R.

Note that a peripheral that actually uses AS, has no access to the multiplexed data/address lower order port on the processor, those don't appear on the M8 bus. So the best place for that kind of peripheral is on the same board, it's how I plan to add the RTC. The same RTC could reside on a peripheral card, and use AS from the bus for timing, but it would also have to multiplex the lower order 8 data/address lines. Luckily most MC68xx/63xx peripherals and others, don't use that kind of multiplexing. The RTC is actually the only one I know of.

I was thinking of using a 74xx00 NAND gate to develop WE* and OE*, the same as used on the Searle/Tranter design and others. However, it may be possible to use a 74xx139. Surprisingly, the 139 is not as common in stock as most other 74 series. A few local places didn't have it, I had to go across town 15km. The existing decode circuit is compact and efficient, but not very scalable. I used the '139 because it filled a "sweet spot" to allow only two IC's to perform all the decoding. The 512k by 8 SRAM requires some additional address logic, a second, memory mapped 74xx373 to latch A16-A18 for page selection. In my view, that suggests a complete re-think of the address decoding block.

I consider the existing decode fully functional, produces all the types of select signals that are needed, however simply does not have enough of them. It's worth a try, but an expanded circuit might not be able to leverage the same fundamental logic that permitted the use of only 2 chips. So it's an open question at this point, how exactly to implement it. One thing I am adamant about, is no use of GAL's or PAL's that need to be programmed.

Honestly, I meant to spend some time thinking about the RAM paging method, but a long weekend and other factors delayed it. That is at the heart of how the memory timing/decoding circuit has to be made.

**Comment by KenWillmott** [comment]
I write this separately because it's unrelated, for a long time I've wondered whether RD* and WR* should be added to the bus. Basically, this would involve adding them to four extra pins on the "end" of the existing 40 pin interface. This would preserve the compatibility of 20, 40, and 44 pin expansion boards. Then it would never be necessary to duplicate them on expansion boards when they are needed. But it might propel the most common interface towards the 44 pins, since boards designed for it wouldn't work on a 40 pin main board interface. I'd rather avoid it. They can always be developed from E and R/W* using the NAND gate method.

At the end of the road, I'd like to see it possible to mix and match projects, for example using the game board with a 6502 main board or something like that. The M8 should maximize inter-operability between different processors and peripherals in the same general family (Motorola style synchronous 8 bit bus).

**Comment by TheEccentricGenius** [comment]
AS will go onto to pin 29, I still hadn't put it anywhere in the bus schematic as I wanted your input.

I will put a pause on the development till you get back about the memory mapping.

Last night I was thinking, _Just chunking a GAL or PAL in would make things easy_ but no I agree whole heartedly. It would just be a pain long term.

Don't mind the NAND gate, but I'm still toying with the RTC clock idea.... Probably good overall.

I just had the same layout as the 63C03Y board.... Struggling with NVRAM like you were.

**Comment by KenWillmott** [comment]
The easy alternative to the HD146818 is to just provide an additional serial port with an HD63B50 or MC68B50. It's extremely tempting, could support a co-processor like the ESP32, provide RTC, keypad, OLED, things like that, or even NTP time via wifi. It's very generic and could support a variety of devices, anything that has serial...

**Reply by TheEccentricGenius** [reply]
You've won me on the serial chip, will be even harder to convince oneself to just use a RTC chip.

Will the EEPROM be first in the address decoding block?
I am struggling to see how to do RAM paging with one NVRAM chip, but will wait till you've sorted it out I guess.

**Comment by KenWillmott** [comment]
Here is the basic layout, the paging is done by the upper latch, which is memory mapped. The upper order address values are written to it, then in this example A16-18 decode for the NVRAMs, A19 goes to the decoder to select one NVRAM (thus total memory is 1Mb).
![image](https://github.com/KenWillmott/HD6303EVB/assets/17345651/8444b861-bb5c-47a4-9635-67c560f503ec)
The EEPROM mapping is still just a question in my mind at this point, I'm not sure. There doesn't have to be more than one NVRAM IC, but I showed two, as a demonstration.

Something important I overlooked... the 6303R has no MR pin, this means that clock cycle stretching is not possible; the bus must run at the maximum speed of the slowest I/O peripheral, or less. With a 7.3728MHz crystal, this means only 2MHz peripheral chips (usually "B" rated parts) will work. This automatically rules out the RTC except at lower CPU clock speeds.

**Reply by TheEccentricGenius** [reply]
I'll only be getting one NVRAM ic, to keep down on project costs.
Thanks for the updated map will rearrange my circuit to match.

I think that clarifies on the RTC, will you put on HD63B50 or have it on a expansion board?

**Comment by KenWillmott** [comment]
I should mention one wrinkle. The 74xx373 has no reset function, outputs may be random until it's written to. So the paging latch has to be initialized before any external RAM is used.  It will work, but needs attention at boot up. The 74xx374 or 74xx273 has reset which could be connected to system reset. However, tri-state control on the latch is useful if you plan to support bus arbitration from the expansion connector.

Edit - I previously was confused about the clock polarity of the upper address latch. The 373 latches on a low going edge, which is perfect for the lower latch that is strobed by AS. But although data is normally clocked on falling E, the select signals are of opposite polarity to E, so the latched data should be clocked by a rising edge (at the end of a bus cycle). This makes a 374 or 273 a more obvious choice. For the 74xx373 based circuit posted above to work, LATSEL would have to be active high.

Yes, I will include a second serial port on the main board, so it can share a baud clock with the on chip UART, and because it is so generally useful. To be honest, I'm now leaning towards remaining with the 6303Y, but that won't change much about the design. I decided to bite the bullet and move to a 4 layer PCB for any main board. Peripheral boards like the game board, I will try to do on 2 sides if I can to save cost.

I'm waiting for SAA-1099 sound chip to arrive for testing before starting work on the game board layout.

**Reply by TheEccentricGenius** [reply]
That probably won't be a problem but I guess we will see.

I'll be getting the serial regardless if it is included or not, but definitely is a useful chip.
 I'll agree the 6303Y is looking a bit more attractive of a proposition now.

I'll stick to the 74xx373 for now(in case I need bus arbitration).

**Comment by KenWillmott** [comment]
I'm looking into piggybacking a 512k RAM into the existing socket. Then I can test it pretty quickly. No changes to the memory map in that case. Just one select line for the extended address latch used up, at the moment.

**Comment by KenWillmott** [comment]
Ok, that experiment seems a success. I installed a 512k by 8 RAM in the NVRAM socket, no changes to the EVB, except one fly wire to pick up the select signal from IOSEL*. I used a 273, with reset unused. So any of the latch or octal FF IC's will work, provided the clock polarity is positive going. If you want tristate capability, that would be a 74xx374 but you can also use the 74xx574 which has organized all the I/O pins to be grouped together if it makes the layout easier. In this case, I'm using D0-D4 mapped to A14-A18 on the SRAM, which produces 32 pages, each page 16kb in size. A 74xx373 will work if the clock polarity is reversed. Sometimes that comes for free if you have gated the select with some NAND or NOR that inverts the sense of the logic.

**Reply by TheEccentricGenius** [reply]
So are you going to change memory mapping or not?

I've completed my order the components will coming within a month(I hope sooner), was going to order the 273, but found ten for another project I was doing.

I'll only put the 373 in if I have too,

How are other tests been going....?

**Comment by KenWillmott** [comment]
Those are good questions. It hinges a bit on the road map - because I anticipate it branching into different boards. I will start a road map discussion for that, I already made a tentative graphic showing different paths. The current design is really supposed to be just an upgrade and bug fix for the existing EVB. So the memory map and decode circuit won't change. Only, I will try to gate the latch address so it can be shared. If not, I will just drop one expansion connector.

With a mux, you could split one memory IC into paged/non-paged sections, but it's at least one extra IC to do that. So I thought, a 32k RAM is also one IC, by dropping it and adding the paging mux, you haven't really got any parts count reduction. Since there is at least a paging latch, and possibly an on board UART coming, it might be a good idea to tack on some extra decode for those... but those decodes would stay in the same memory area of $D800-$DFFF. It's optional, and depends on how many M8 slots you want to have on board. For this version, I'm leaning towards losing the 20 pin M8 and having just the one 40 pin M8.

I spent about 10 days trying to interface the PCF8584 I2C chip. It was designed when the I2C spec was fresh, and is a complicated chip to operate. I followed the suggested flow charts in the data sheet but ultimately couldn't get it to communicate with the I2C test device. So I set that aside for possible future work. It's challenging.

Anyway, to answer your main question, no, I'm not thinking of changing the memory map right now. The proposed paging circuit consists of no more than the latch, conveying the low order 5 data bits D0-D4 to the upper 5 address bits of the 512kb NVRAM. I want to reserve 2 of the remaining latch bits for memory expansion, so D5-D6 to support a maximum 2Mb of memory, D7 reserved for unspecified future expansion. D5 is especially important in that regard, since some variant may have a single AS6-8008 1Mb NVRAM.

However, total 544kb RAM is nothing to sneeze at!

**Reply by TheEccentricGenius** [reply]
`However, total 544kb RAM is nothing to sneeze at!`
Agreed.

**Comment by TheEccentricGenius** [comment]
So there is an 512k NVRAM chip, 32k RAM chip and an EEPROM chip.
With the serial chip possibly included. 
Not bad at all.

I don't know about the 20pin M8, I was always dubious of it....But it is your design and I could see a few ways it could be used.

Just glanced at the PCF8584 datasheet, I have to agree with the challenging aspect. But definitely would be nice to have that onboard.

I'll look out for the discussion.

**Comment by KenWillmott** [comment]
About the 20 pin, the M8 philosophy is to provide as many features that are cheap or free, as possible, but non-critical aspects are optional. So the number of slots, whether they are 20 or 40 pin, is a board designer's whim. The only mandatory thing at the moment, is that a peripheral board X will plug into a controller (CPU) board Y, and it will work. The idea for the 20 pin actually came from the Apple II bus. However there are too many factors to define it completely, so things like bus speed and buffering are also left open. It's possible that I might define benchmarks for those later on, for now the motto is "just be reasonable".

I can send you the non-working PCF8584 code I wrote, if you want to play with it. It does produce activity on the I2C bus, however it only sends a start pulse, no data. I have no idea where I'm going wrong.

**Reply by TheEccentricGenius** [reply]
I don't currently have a PCF8584 chip, send the code if you want as I may buy one.... 
_They may be useful for other projects..._

`"just be reasonable".`
sounds reasonable...Pardon the pun :)

**Comment by KenWillmott** [comment]
It was too hasty. By ''reasonable" I just mean when interfacing the bus, not do shortcuts that defeat the potential of some signal line. For example, I jumpered the I/O select space to MR on the bus, this worked with no boards plugged in, but it defeats the use of MR on the bus. It is only temporary. When a conflict or question in the designers mind comes up, it is better to open a discussion or investigation of some sort, to see what can be done.

It reminds me that I should make MR an issue, to define it as a common drain and add a pull up resistor on the CPU board.
